compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,851 @@
+  std::map<Type*, int64_t> typeMap;
+  fileinfo fiMarshalling;
+  fileinfo fiClientBundle;
+  fileinfo fiServerBundle;
+  GenInfo* info;
+
+  bool shouldEmit(ModuleSymbol* md);
+  bool shouldEmit(FnSymbol* fn);
+  void setOutput(fileinfo* fi);
+  void setOutputAndWrite(fileinfo* fi, const std::string& gen);
+  void write(const std::string& code);
+  int64_t assignUniqueTypeID(Type* t);
+  void emitClientWrapper(FnSymbol* fn);
+  void emitServerWrapper(FnSymbol* fn);
+  bool isSupportedType(Type* t);
+  void verifyPrototype(FnSymbol* fn);
+  Type* getTypeFromFormal(ArgSymbol* as);
+  Type* getTypeFromFormal(FnSymbol* fn, int i);
+  bool typeRequiresAllocation(Type* t);
+
+  std::string genMarshalBodyPrimitiveScalar(Type* t, bool out);
+  std::string genComment(const char* msg, const char* pfx="");
+  std::string genNote(const char* msg);
+  std::string genTodo(const char* msg);
+  std::string genDefine(const char* val);
+  std::string genHeaderInc(const char* header, bool system=false);
+  std::string genMarshalRoutine(Type* t, bool out);
+  std::string genMarshalPushRoutine(Type* t);
+  std::string genMarshalPullRoutine(Type* t);
+  std::string genServerDispatchSwitch(const std::vector<FnSymbol*>& fns);
+  std::string genDebugPrintCall(FnSymbol* fn, const char* pfx="");
+  std::string genDebugPrintCall(const char* msg, const char* pfx="");
+  std::string genFuncNumericID(FnSymbol* fn);
+  std::string genServerWrapperCall(FnSymbol* fn);
+  std::string genClientsideRPC(FnSymbol* fn);
+  std::string genServersideRPC(FnSymbol* fn);
+  std::string genMarshalCall(const char* skt, const char* var, Type* t,
+                             bool out);
+  std::string genMarshalPushCall(const char* skt, const char* var, Type* t);
+  std::string genMarshalPullCall(const char* skt, const char* var, Type* t);
+  std::string genTypeName(Type* t);
+  std::string genSocketCall(const char* skt, const char* var, const char* len,
+                            bool out);
+  std::string genSocketCall(const char* skt, const char* var, bool out);
+  std::string genSocketPushCall(const char* skt, const char* var);
+  std::string genSocketPullCall(const char* skt, const char* var);
+  std::string genAddressOf(const char* var);
+  std::string genAddressOf(std::string& var);
+  std::string genSizeof(const char* var);
+  std::string genSizeof(std::string& var);
+  std::string genNewDecl(const char* t, const char* n);
+  std::string genNewDecl(Type* t, const char* n);
+
+};
+
+//
+// Generic helper method replacing C++11 std::to_string() method.
+//
+template <typename T>
+std::string str(T value) {
+  std::ostringstream tmp;
+  tmp << value;
+  return tmp.str();
+}
+
+//
+// This is the main entrypoint for MLI code generation, call this if the
+// necessary conditions are met in codegen().
+//
+void codegenMultiLocaleInteropWrappers(void) {
+  Vec<ModuleSymbol*> &mds = allModules;
+
+  // Insert all kinds of debug printlines into generated code for now.
+  MLIContext mli(true);
```

Comment:
I forget, do you have an issue to tracking turning this on/off via a developer compiler flag?  If not, I would add one - I would expect us to want it turned off before we run testing with these changes (and definitely before advertising this feature in a release - users will likely not want this kind of debugging output inserted into their program for them.  It also would expose implementation details to the user, which is bad practice)

Reply:
I can just turn it off for now (but yes, opening an issue is on a TODO).

Reply:
Thanks for reminding me.